transaction/std/src/transaction_builder.rs

```diff
@@ -246,42 +247,212 @@ impl<FPR: FogPubkeyResolver> TransactionBuilder<FPR> {
+    /// * validate it (call .validate())
+    /// * check if key image appeared already (call .key_image())
+    /// * provide merkle proofs of membership for each ring member (see
+    ///   .tx_out_global_indices)
+    ///
+    /// # Arguments
+    /// * `signed_contingent_input` - The pre-signed input we are adding
+    /// * `sci_change_amount` - The amount of value of the SCI which we are
+    ///   returning to the signer. This determines the "fill fraction" for the
+    ///   partial-fill rules. This will be equal to the
+    ///   real_change_output_amount.
+    ///
+    /// # Returns
+    /// * A list of all outlay amounts deduced to fulfill the fractional output
+    ///   rules, in the cheapest way possible.
+    pub fn add_presigned_partial_fill_input(
+        &mut self,
+        sci: SignedContingentInput,
+        sci_change_amount: Amount,
+    ) -> Result<Vec<Amount>, SignedContingentInputError> {
+        if !self.block_version.masked_amount_v2_is_supported() {
+            return Err(
+                SignedContingentInputError::FeatureNotSupportedAtBlockVersion(
+                    *self.block_version,
+                    "partial fills",
+                ),
+            );
+        }
         // Check if the sci already has membership proofs, the caller is supposed to do
         // that
         if sci.tx_in.ring.len() != sci.tx_in.proofs.len() {
             return Err(SignedContingentInputError::MissingProofs);
         }
-        if let Some(rules) = sci.tx_in.input_rules.as_ref() {
-            // Enforce all rules so that our transaction will be valid
-            if rules.required_outputs.len() != sci.required_output_amounts.len() {
-                return Err(SignedContingentInputError::WrongNumberOfRequiredOutputAmounts);
-            }
-            // 1. Required outputs
-            for (required_output, unmasked_amount) in rules
-                .required_outputs
+        let rules = sci
+            .tx_in
+            .input_rules
+            .as_ref()
+            .ok_or(SignedContingentInputError::MissingRules)?;
+        let partial_fill_change = rules
+            .partial_fill_change
+            .as_ref()
+            .ok_or(SignedContingentInputError::MissingPartialFillChange)?;
+        let (partial_fill_change_amount, partial_fill_change_blinding) =
+            partial_fill_change.reveal_amount()?;
+        if partial_fill_change_amount.value == 0 {
+            return Err(SignedContingentInputError::ZeroPartialFillChange);
+        }
+        if rules.min_partial_fill_value > partial_fill_change_amount.value {
+            return Err(SignedContingentInputError::MinPartialFillValueExceedsPartialChange);
+        }
+        if partial_fill_change_amount.token_id != sci_change_amount.token_id {
+            return Err(SignedContingentInputError::TokenIdMismatch);
+        }
+        if partial_fill_change_amount.value - rules.min_partial_fill_value < sci_change_amount.value
```

Comment:
Sorry, could you clarify what this check is about? Why are we verifying that the sci_change_amount is greater than the difference between the partial_fill_change_amount and the minimum partial fill value?

Reply:
We have to enforce the `min_partial_fill_value` constraint, or the Tx will be invalid.

The amount that we are attempting to take, out of the partial fill change, is `partial_fill_change_amount.value - sci_change_amount.value`.

The minimum that we must take is `rules.min_partial_fill_value`.

If `partial_fill_change_amount.value - sci_change_amount.value < rules.min_partial_fill_value`, then we broke the rule and that's supposed to be an error. The error status is currently called "change limit exceeded".

This is the same as what we are checking in the code, it's just that we re-balance the equation slightly. That's mainly because when we subtract we need to do a bounds check for negative numbers or we can panic. We earlier tested that min_partial_fill_value <= partial_fill_change_amount, so that's the expression I want to use to do the test.

Reply:
added a comment